### PR TITLE
Store: add jetpack/media-source sub-state tree

### PR DESCRIPTION
**Pull request by retrofox**
This PR register a new redux store in the app, which will be used to put all media components together, providing a communication channel. Also, it integrates with the audio player component in the podcast player block.
I've added a basic README.md file with useful doc.

Fixes #

#### Changes proposed in this Pull Request:

* store: add jetpack/media-source sub-state tree

#### Jetpack product discussion

p1608061925102200?thread_ts=1608055913.079400&cid=CKZHG0QCR-slack-

#### Does this pull request change what data or activity we track or use?

No

### Dispatching actions

* Go to the editor
* Open the client dev console
* Register a media source, defining and ID:

```es6
wp.data.dispatch( 'jetpack/media-source' ).registerMediaSource( 'myMediaSource' );
```

You should see the subtree populated, using the Redux dev tool:

Action | Subtree
-------|--------
image | image

By default, the media source will be populated with the `is-paused` status, and position `0`. Pass a second argument in case you'd like to set the media source status.

```es6
wp.data.dispatch( 'jetpack/media-source' ).registerMediaSource( 'myMediaSource', {
  status: 'is-playing',
  position: 100,
} );
```

Sidenote: we're going to define the playing status with constant, already defined [here](https://github.com/Automattic/jetpack/pull/17879).

### Podcast player

When a podcast player instance is inserted in the editor, it registers the media source. Confirm this by creating a new instance, keeping in mind that it will be registered only when the audio gets a valid URL.
Feel free to use this podcast URL: https://distributed.blog/2020/10/15/episode-25-davit-baghdasaryan-on-the-science-of-sound-in-a-distributed-work-world/

* Podcast block inserted with a valid URL
image

* media source registered in the subtree
Screen Shot 2020-12-18 at 2 54 08 PM

The id is defined by the id attribute of the media element: `mejs_4254545776144716_html5`

* Remove the podcast player, it should unregister the media source.

Action | State tree
-------|-------
image | image

#### Proposed changelog entry for your changes:

* store: add jetpack/media-source sub-state tree

### CODE REVIEW

**Comment by allilevine** [comment]
@retrofox I'm able to register and unregister the players as expected, but will these changes need to be refactored when #17879 is merged?

**Comment by retrofox** [comment]
> @retrofox I'm able to register and unregister the players as expected, but will these changes need to be refactored when #17879 is merged?

I'd say update instead of refact. Provably we'd like to register the media source from the podcast player block instead of the audio player.  But yeah, definitely something will change when these PRs land.

File: extensions/shared/components/audio-player/index.jsx [hunk 1]
```diff
@@ -72,8 +74,15 @@ function AudioPlayer( {
 
 	useEffect( () => {
 		const audio = audioRef.current;
-		// Initialize MediaElement.js.
```
Comment (allilevine) [inline extensions/shared/components/audio-player/index.jsx:75]: Does using the store here resolve the issue with `MediaElement.js` described [here](https://github.com/Automattic/jetpack/blob/master/extensions/shared/components/audio-player/README.md#potential-issues)? 

> Currently the component is a wrapper around a MediaElement.js component which does its own event handling. It's possible therefore, that the state of the player can get out of sync with the state of the controlling component. We have seen an example of this caused by rapid clicking on the track progress bar, which was solved by debouncing the events.

Comment (retrofox) [reply extensions/shared/components/audio-player/index.jsx:75]: Not sure whether it could be related. Honestly, I haven't seen these _out of sync_ issues while testing and developing.

File: extensions/shared/components/audio-player/index.jsx [hunk 2]
```diff
@@ -105,6 +114,7 @@ function AudioPlayer( {
 			onPlay && audio.removeEventListener( 'play', onPlay );
 			onPause && audio.removeEventListener( 'pause', onPause );
 			onError && audio.removeEventListener( 'error', onError );
```
Comment (allilevine) [inline extensions/shared/components/audio-player/index.jsx:116]: Is the next step to add store actions for playing and pausing?

Comment (retrofox) [reply extensions/shared/components/audio-player/index.jsx:116]: yes, everything should be controlled by the state tree. Also setting the current time position, among other ones.

File: extensions/shared/components/audio-player/index.jsx (hunk 1 above)
Comment (allilevine) [reply extensions/shared/components/audio-player/index.jsx:75]: Thanks! Should the `// Initialize MediaElement.js.` comment stay for now?

Comment (retrofox) [reply extensions/shared/components/audio-player/index.jsx:75]: yes, as long as we create a new MediaElement instance we should comment that. We don't have planned to remove the MediaElement in the short time AFAIK.

**Comment by mreishus** [comment]
Added unit tests, run `yarn run test-extensions`
[update/add-media-state-tree 3ee2aa3e4] store/media-source: Add unit tests

File: extensions/store/media-source/storeDefinition.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+					},
+				};
+
+			case 'UNREGISTER_MEDIA_SOURCE':
+				// eslint-disable-next-line no-case-declarations
+				const currentState = Object.assign( {}, state );
+				if ( currentState.players[ action.id ] ) {
+					delete currentState.players[ action.id ];
+				}
+				return currentState;
```
Comment (mreishus) [inline extensions/store/media-source/storeDefinition.js:47]:
```suggestion
			case 'UNREGISTER_MEDIA_SOURCE': {
				const currentState = Object.assign( {}, state );
				if ( currentState.players[ action.id ] ) {
					delete currentState.players[ action.id ];
				}
				return currentState;
			}				
```
Adding brackets should fix the lint error and prevent future conflicts if other action types have a `currentState`.

**Review by mreishus — DISMISSED** [review]
I may have broken linting with the tests I added. I see an error about 'expect' is not defined here: https://github.com/Automattic/jetpack/actions/runs/478394109. However, other files like this one use expect() without importing it. Anyone know how to fix?

> We might have to ask Jeremy?

yes, I think so 😅  cc @jeherve

**Comment by retrofox** [comment]
> I may have broken linting with the tests I added. I see an error about 'expect' is not defined here: https://github.com/Automattic/jetpack/actions/runs/478394109. However, other files like this one use expect() without importing it. Anyone know how to fix?

I think it will be fixed by 515fafcfefa8711ef8f3b271a107fd896caa1199

**Review by mreishus — APPROVED** [review]
LGTM :heavy_check_mark:

File: extensions/store/media-source/test/.eslintrc.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
Comment (jeherve) [inline extensions/store/media-source/test/.eslintrc.js:1]: What if you were to apply this to `extensions/.eslintrc.js` directly, so we don't run into the issue again in the future? I'm actually surprised this hasn't come up in the past.

File: extensions/editor.js [hunk 5]
```diff
@@ -10,6 +10,8 @@ import './extended-blocks/core-embed';
 import './extended-blocks/paid-blocks';
 import './shared/styles/slideshow-fix.scss';
 
+import './store';
```
Comment (jeherve) [inline extensions/editor.js:13]: Maybe we should add a comment here, so anyone looking at this main file quickly understands what the store can be used for?

File: extensions/shared/components/audio-player/index.jsx [hunk 6]
```diff
@@ -72,8 +74,15 @@ function AudioPlayer( {
...
-		const mediaElement = new MediaElementPlayer( audio, meJsSettings );
+		const mediaElement = new MediaElementPlayer( audio, {
+			...meJsSettings,
+			success: function() {
+				dispatch( STORE_ID ).registerMediaSource( audio.id, {
+					status: 'is-paused',
+					timestamp: 0,
+				} );
+			}
+		} );
```
Comment (jeherve) [inline extensions/shared/components/audio-player/index.jsx:85]: Does this mean we're only taking the initial, default tracks into account? If I extend the number of tracks later on via the block settings, should a new media source be added to the source tree for each one of the new tracks?

**Review by jeherve — CHANGES_REQUESTED** [review]

File: extensions/shared/components/audio-player/index.jsx (hunk 6 above)
Comment (retrofox) [reply extensions/shared/components/audio-player/index.jsx:85]: It's a good question. From my point of view, should exist a media-source register per audio player. Right now, when you add a podcast player, it inserts only one `<audio>` element. The block has the ability to switch among different sources (tracks), but the audio instance is always the same.
So the answer is not, it should not register a new media-source for each track but for the podcast player.

Something to keep in mind: This is just the initial step of the global state implementation. We already have follow-up PRs with significant changes. However, this PR provides the initial structure.

File: extensions/editor.js (hunk 5 above)
Comment (retrofox) [reply extensions/editor.js:13]: can we combine different events there?
We should set `env: { jest: true },`...

File: projects/plugins/jetpack/extensions/store/media-source/README.md [hunk 7]
```diff
@@ -0,0 +1,53 @@
+Media Source State Tree
+=======================
+
+This subtree takes over registering media sources globally in the editor app context. Initially created with the purpose to put all components together of the anchor.fm integration provides a communication channel allowing controlling, for instance, the audio state from `play` to `stop`, and vice versa.
```
Comment (pablinos) [inline projects/plugins/jetpack/extensions/store/media-source/README.md:4]: > Initially created with the purpose to put all components together of the anchor.fm integration provides a 

Maybe we could rephrase this to something like:

Initially created for the anchor.fm integration, it provides a communication channel for controlling, for instance, ...

**Review by pablinos — APPROVED** [review]
This works well for me. I can register players and can see them being registered by the podcast block. The play status isn't changing as far as I can see, but that's probably better as part of the other PRs

I've tested it in the editor and the frontend both in a Jetpack site and on WordPress.com

**Comment by retrofox** [comment]
> This works well for me. I can register players and can see them being registered by the podcast block. The play status isn't changing as far as I can see, but that's probably better as part of the other PRs

That's right. There is another PR (linked here) waiting for this one which takes over of the full handling. Thanks!
> 
> I've tested it in the editor and the frontend both in a Jetpack site and on WordPress.com

File: projects/plugins/jetpack/extensions/store/media-source/README.md (hunk 7 above)
Comment (retrofox) [reply projects/plugins/jetpack/extensions/store/media-source/README.md]: I've changed the description, but it's something we'd like to keep updated with the API, so probably going to change soon.

**Review by pablinos — APPROVED** [review]
LGTM!

**Review by jeherve — APPROVED** [review]

**Comment by retrofox** [comment]
r219477-wpcom. Thanks, folks/
